### PR TITLE
Updated the restoreCase logic to check for upper and lower case pluralization

### PR DESCRIPTION
**Pull request by lappang-cheung**
I'd updated the logic for the restoreCase as it has missed some testcases as well and it shall pass for the new ones. And this is the solution for [https://github.com/blakeembrey/pluralize/issues/131](url)

```
it('should allow new plural matching rules', function () {
      expect(pluralize.plural('vIP')).to.equal('vIPS');

      pluralize.addPluralRule(/vIvP$/i, 'vIvPs');

      expect(pluralize.plural('vIvP')).to.equal('vIvPs');
    });

    it('should allow new plural matching rules', function () {
      expect(pluralize.plural('IP')).to.equal('IPS');

      pluralize.addPluralRule(/IP$/i, 'IPs');

      expect(pluralize.plural('IP')).to.equal('IPs');
    });

    it('should allow new plural matching rules', function () {
      expect(pluralize.plural('Blank Application2')).to.equal('Blank Application2s');

      pluralize.addPluralRule(/tion2$/i, 'tion2S');

      expect(pluralize.plural('Blank Application2')).to.equal('Blank Application2S');
    });
```

Screen Shot 2019-11-01 at 10 44 06 AM

### CODE REVIEW

File: pluralize.js [hunk 1]
```diff
@@ -49,6 +49,11 @@
     // Tokens are an exact match.
     if (word === token) return token;
 
+    // Starts with the same letters
+    if (token.toLowerCase().startsWith(word.toLowerCase())) {
```
Comment (blakeembrey) [inline pluralize.js]: This is a breaking change for supported environments of `pluralize`. We should change `startsWith` to something else. Can you also remove the token logic below? This probably overlaps in implementation.

File: .gitignore [hunk 2]
```diff
@@ -3,3 +3,5 @@ node_modules
 components
 build
 coverage
+yarn.lock
```
Comment (blakeembrey) [inline .gitignore]: Please don't change this.

**Comment by lappang-cheung** [comment]
Hey Blake,

I'd undid the changes to the ".gitignore" file and updated the "pluralize" file.

As for my logic: 

```
// This is an unique case, where we try to preserved the words that aren't changed significantly.
    // E.g. "sTaRTREK" (plural form is sTaRTREKs).
    if ((token.toLowerCase().substring(0, word.length)) === word.toLowerCase()) {
      return word + token.substring(word.length);
    }

    // Lower cased words. E.g. "hello".
    if (word === word.toLowerCase()) return token.toLowerCase();

    // Upper cased words. E.g. "WHISKY".
    if (word === word.toUpperCase()) return token.toUpperCase();
```
This parts only check for special cases where we have weird combination when calling "pluralize" method and if this checks fails then it goes to the base case. Such as "sTaRTREK" and wanting the result to be "sTaRTREK

Screen Shot 2019-11-01 at 1 57 48 PM
S".

File: .gitignore (hunk 2 above)
Comment (lappang-cheung) [reply .gitignore]: Updated the code with:

```
if ((token.toLowerCase().substring(0, word.length)) === word.toLowerCase()) {
      return word + token.substring(word.length);
    }
```

File: pluralize.js (hunk 1 above)
Comment (lappang-cheung) [reply pluralize.js]: Updated the code with:

```
if ((token.toLowerCase().substring(0, word.length)) === word.toLowerCase()) {
      return word + token.substring(word.length);
    }
```

These parts only check for special cases where we have weird combinations when calling the "pluralize" method and if this check fails then it goes to the base case. Such as "sTaRTREK" and wanting the result to be "sTaRTREKS".

**Comment by blakeembrey** [comment]
I’d like to see more tests and unnecessary changes reverted before I can merge it. Based on this PR, I’m pretty sure it’ll break anyone using upper cased words. At that point, it’s probably better to just always use lowercase and not have this logic at all.

**Comment by blakeembrey** [comment]
Actually, am I understanding this correctly that it'll _only_ word if the entire token is matched? In that case, I think it'd be better that we find a more appropriate solution. It also doesn't solve https://github.com/blakeembrey/pluralize/issues/131 unless you intend to add manual overrides to every word, which seems redundant.

**Comment by blakeembrey** [comment]
If this is a real issue for people, I propose just removing the feature altogether and making the consumers of the library handle the casing they want. It'd make it slightly faster, make those overrides "expected", _and_ making it to upper case is easier than doing all the other magic.

**Comment by blakeembrey** [comment]
Here's a test case to consider that fails with this change, if you're planning to add more tests:

```
> pluralize.plural('SCHEMA')
'SCHEMAta'
```

**Pull request closed without merging**